### PR TITLE
Refactor/node names

### PR DESCRIPTION
**Pull request by Yadunund**
## Refactor

What started out as a simple fix for appropriately renaming the ROS 2 nodes and executable (#2) turned out to be a complete refactor of `rmf_schedule_visualizer`.

Summary of changes:
* `src/VisualizerData.hpp` renamed to `ScheduleDataNode.hpp` and moved into public API while adopting PIMPL pattern
* `src/Server.hpp` renamed to `TrajectoryServer.hpp` and moved into public API while adopting PIMPL pattern
* Exportable `rmf_schedule_visualizer` library of public API
* Single `main.cpp` which publishes RVIZ markers for schedule visualization and starts the websocket server to serve trajectories
* Renamed ROS 2 executable from `rviz2` to `schedule_visualizer`
* Renamed nodes `/viz` and `rviz_node` to `rmf_schedule_visaulizer_data_node` and `rmf_schedule_visualizer_marker_publisher`
* Parse `headless` argument in `visualizer.launch.xml` to determine whether to launch `RViz`

### CODE REVIEW

**Comment by Yadunund** [comment]
@koonpeng could you please test this branch to make sure nothing is broken for the dashboard?
@ddengster could you also test this branch to make sure nothing is broken with the negotiation viz/server?

**Comment by Yadunund** [comment]
This branch is required for testing with `rmf_demos`
https://github.com/open-rmf/rmf_demos/pull/11

**Comment by ddengster** [comment]
> @ddengster could you also test this branch to make sure nothing is broken with the negotiation viz/server?

Doesnt seem like anything's changed on this backend server, should be GTG

File: rmf_schedule_visualizer/include/rmf_schedule_visualizer/ScheduleDataNode.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  /// Get a mutable reference to the mutex used by the ScheduleDataNode
+  std::mutex& get_mutex();
+
+  /// Get a shared pointer to the ROS 2 negotiation interface
+  std::shared_ptr<Negotiation> get_negotiation();
+
+  class Implementation;
+
+private:
+  ScheduleDataNode(std::string _node_name);
```
Comment (gbiggs) [inline rmf_schedule_visualizer/include/rmf_schedule_visualizer/ScheduleDataNode.hpp:85]: I won't refuse an approval over this, but in general the node should decide its name, not the function that constructs it. If the node name needs to be changed remapping can be used.

A node name is a data type name, not a variable name. The node instance name is the variable name and remapping is how you name the variable.

File: rmf_schedule_visualizer/src/main.cpp [hunk 2]
```diff
@@ -15,11 +15,796 @@
...
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "nav_msgs/msg/occupancy_grid.hpp"
+
+using namespace std::chrono_literals;
+
+//==============================================================================
+class ScheduleMarkerPublisher : public rclcpp::Node
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/main.cpp:53]: This is huge. Consider moving it to a separate file.

File: rmf_schedule_visualizer/src/main.cpp [hunk 3]
```diff
@@ -97,13 +891,25 @@ int main(int argc, char* argv[])
...
     "Websocket server started on port: " + std::to_string(port));
+  
+  auto rviz_node = std::make_shared<ScheduleMarkerPublisher>(
+    "rmf_schedule_visualizer_marker_publisher",
+    schedule_data_node,
+    std::move(map_name),
+    rate);
+
+  rclcpp::executors::MultiThreadedExecutor executor{
+    rclcpp::executor::ExecutorArgs(), 3
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/main.cpp:904]: Why 3?

File: rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+#include <rmf_schedule_visualizer/ScheduleDataNode.hpp>
+
+#include <rmf_traffic_ros2/StandardNames.hpp>
+#include <rmf_traffic_ros2/Time.hpp>
+#include <rmf_traffic_ros2/Trajectory.hpp>
+#include <rmf_traffic_ros2/schedule/MirrorManager.hpp>
+
+#include <rmf_traffic_msgs/msg/negotiation_notice.hpp>
+#include <rmf_traffic_msgs/msg/negotiation_conclusion.hpp>
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp:26]: Minor nitpick: Place includes in alphabetical order.

File: rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+  auto mirror_mgr_future = rmf_traffic_ros2::schedule::make_mirror(
+    *schedule_data, rmf_traffic::schedule::query_all(),
+    &schedule_data->_pimpl->mutex);
+
+  const auto stop_time = start_time + wait_time;
+  while (rclcpp::ok() && std::chrono::steady_clock::now() < stop_time)
+  {
+    rclcpp::spin_some(schedule_data);
+    using namespace std::chrono_literals;
+    bool ready = (mirror_mgr_future.wait_for(0s) == std::future_status::ready);
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp:113]: Why is the result of this comparison stored in a variable that's only used in the very next line?

File: rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+  rmf_traffic::Duration wait_time)
+{
+  using ConflictNotice = rmf_traffic_msgs::msg::NegotiationNotice;
+  using ConflictConclusion = rmf_traffic_msgs::msg::NegotiationConclusion;
+
+  const auto start_time = std::chrono::steady_clock::now();
+  std::shared_ptr<ScheduleDataNode> schedule_data(
+    new ScheduleDataNode(std::move(node_name)));
+
+  schedule_data->_pimpl->conflict_notice_sub =
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp:81]: Consider moving some of the `Implementation` setup into that class to shorten this function.

File: rmf_schedule_visualizer/src/rmf_schedule_visualizer/TrajectoryServer.cpp [hunk 7]
```diff
@@ -377,19 +350,129 @@ std::string Server::parse_trajectories(
   return response;
 }
 
-Server::~Server()
+//==============================================================================
+std::shared_ptr<TrajectoryServer> TrajectoryServer::make(
```
Comment (gbiggs) [inline rmf_schedule_visualizer/src/rmf_schedule_visualizer/TrajectoryServer.cpp:354]: Consider moving some of this inside the pimpl object to shorten and simplify this function.

**Review by gbiggs — COMMENTED** [review]
In general it's good, but I have a couple of comments I'd like you to look at before I hit approve.

File: rmf_schedule_visualizer/include/rmf_schedule_visualizer/ScheduleDataNode.hpp (hunk 1 above)
Comment (Yadunund) [reply rmf_schedule_visualizer/include/rmf_schedule_visualizer/ScheduleDataNode.hpp]: Thanks for the heads up. Will keep this in mind in the future

File: rmf_schedule_visualizer/src/main.cpp (hunk 3 above)
Comment (Yadunund) [reply rmf_schedule_visualizer/src/main.cpp]: Fixed. This should be 2. Thanks for pointing it out.

**Review by gbiggs — APPROVED** [review]

**Pull request merged**